### PR TITLE
Fixed bug where Segment.GetCursor() would throw an ArgmentOutOfRange Exception if the Segment has no Shards

### LINKED ISSUE #20950: [BUG] Sporadic ArgumentOutOfRangeException when calling BlobChangeFeedClient.GetChangesAsync
**Query/Question**
The following exception gets thrown sporadically when executing the BlobChangeFeedClient.GetChangesAsync:
`System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')

```
at Azure.Storage.Blobs.ChangeFeed.Segment.GetCursor()
   at Azure.Storage.Blobs.ChangeFeed.ChangeFeed.GetCursor()
   at Azure.Storage.Blobs.ChangeFeed.ChangeFeed.GetPage(Boolean async, Int32 pageSize, CancellationToken cancellationToken)
   at Azure.Storage.Blobs.ChangeFeed.BlobChangeFeedAsyncPageable.AsPages(String continuationToken, Nullable`1 pageSizeHint)+MoveNext()
   at Azure.AsyncPageable`1.GetAsyncEnumerator(CancellationToken cancellationToken)+MoveNext()
   at Azure.AsyncPageable`1.GetAsyncEnumerator(CancellationToken cancellationToken)+MoveNext()
```

We're using the blob change feed nuget package in some integration/automated unit tests to confirm our system is functioning properly. Once the blobs show up (within a set time), the test passes. We're using the change feed because we don't know the exact fullname of the blobs, which get guids appended to the end of the name.

It seems like we run into the issue every few weeks or so (haven't been able to nail down a reproducible path) and usually just re-running the tests causes it to pass.

Here's the code we're using to get the events:
``` c#
BlobChangeFeedClient changeFeedClient = new BlobChangeFeedClient(Configuration.GetConnectionString(GlobalConstants.ArchiveStorageConnectionStringKey));
List<BlobChangeFeedEvent> changeFeedEvents = new List<BlobChangeFeedEvent>();

// IMPORTANT Per Documentation: The change feed client will round start time down to
// the nearest hour, and round endTime up to the next hour if you provide DateTimeOffsets
// with minutes and seconds.

// You can also provide just a start or end time.
await foreach (BlobChangeFeedEvent changeFeedEvent in changeFeedClient.GetChangesAsync(
                start: startTime))
{
  changeFeedEvents.Add(changeFeedEvent);
}

return changeFeedEvents.OrderByDescending(e => e.EventTime).ToList();
```

Is there anything we're doing wrong when getting the changes?

**Environment:**
 - Azure.Storage.Blobs.ChangeFeed 12.0.0-preview.11
 - Executed in VS Tests and run up in Azure Hosted VM during Azure DevOps Release pipeline
 - .NET Core Info
 ```
 .NET Core SDK (reflecting any global.json):
 Version:   3.1.302
 Commit:    41faccf259

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17763
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   D:\_agent\_tool\dotnet\sdk\3.1.302\

Host (useful for support):
  Version: 3.1.6
  Commit:  3acd9b0cd1

.NET Core SDKs installed:
  3.1.302 [D:\_agent\_tool\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.App 3.1.6 [D:\_agent\_tool\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.6 [D:\_agent\_tool\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.6 [D:\_agent\_tool\dotnet\shared\Microsoft.WindowsDesktop.App]
```

 - Windows Server 2019 Version 1809 (OS Build 17763.1879)
 - Visual Studio Enterprise 2019 Version 16.8.4

Please let me know if there's any additional information you need from me.

### PR DESCRIPTION
**Pull request by seanmcc-msft**
Resolves https://github.com/Azure/azure-sdk-for-net/issues/20950

### CODE REVIEW

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Segment.cs [hunk 1]
```diff
@@ -63,7 +63,7 @@ public virtual SegmentCursor GetCursor()
             return new SegmentCursor(
                 segmentPath: ManifestPath,
                 shardCursors: shardCursors,
-                currentShardPath: _shards[_shardIndex].ShardPath);
+                currentShardPath: _shards.Count > 0 ? _shards[_shardIndex].ShardPath : null);
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Segment.cs:66]: Could you add a test where we attempt to deserialize and use cursor with `currentShardPath` = `null` ?
We should make sure we can handle such cursor.

Comment (kasobol-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Segment.cs:66]: we should also make sure other languages can consume such cursor. (and possibly check if they have similar bug..)

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Segment.cs:66]: @gapra-msft @xiafu-msft, can you see if Java and Python have this bug?

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Segment.cs:66]: We can definitely handle `currentShardPath = null`: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/SegmentFactory.cs#L82

Comment (kasobol-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Segment.cs:66]: great, so do we already have test for that edge case ?

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Segment.cs:66]: Done.

**Review by kasobol-msft — APPROVED** [review]

**Pull request merged**